### ISSUE TITLE
Issues using plop on a M1 MacBook Pro

### ISSUE THREAD

**Issue by oakleyarts** [plopjs/plop#298]
I've just installed plop globally on a M1 MacBook Pro, and I'm inside a directory with a plopfile at its root.

As per my plopfile config, I would normally create a new component by running `plop MyComponent`

On my other MacBook it produces a handful of typescript files giving me everything I need to get started.

On this new M1 MacBook Pro I get the following:

`internal/process/esm_loader.js:74
    internalBinding('errors').triggerUncaughtException(
                              ^

Error [ERR_UNSUPPORTED_ESM_URL_SCHEME]: Only file and data URLs are supported by the default ESM loader. Received protocol 'node:'
    at Loader.defaultResolve [as _resolve] (internal/modules/esm/resolve.js:756:11)
    at Loader.resolve (internal/modules/esm/loader.js:85:40)
    at Loader.getModuleJob (internal/modules/esm/loader.js:229:28)
    at ModuleWrap.<anonymous> (internal/modules/esm/module_job.js:51:40)
    at link (internal/modules/esm/module_job.js:50:36) {
  code: 'ERR_UNSUPPORTED_ESM_URL_SCHEME'
}`

Haven't encountered this issue before in my years using plop, so I am _assuming_ its something to do with the M1 differences? Any M1 users out there able to use it as expected?

**Comment by crutchcorn** [comment]
FWIW, we just introduced v3 which changes a lot of things. I suspect that there's a reason for things breaking, but the error message doesn't seem to indicate clearly why. 

Can you share with me your plopfile.js config by any chance? (as well as the name of the file itself and the `type` set in the closest `package.json`?)

**Comment by aminimalanimal** [comment]
@oakleyarts I just received the same error with a non-M1 mac after updating all of my project dependencies to the latest. If you ensure your other Macbook has the same version of node and npm that the new one is using, and trash `package-lock.json` + `node_modules` folder, then `npm i` again, you should be able to reproduce the issue there, too.

Quick fix: `npm i plop@2`

@crutchcorn All I can tell from it is that it's related to ES Modules. I tried replacing all of my `module.exports` with `export default` and `require` with `import` within my configuration files, but that didn't seem to suffice, so I don't know if it's internal to `plop` or not.

> Only file and data URLs are supported by the default ESM loader

**Comment by crutchcorn** [comment]
@aminimalanimal can you share the info requested above? I can't help debug until I see them. 

FWIW I've extensively tested plop with MacOS, even going so far as to write our own E2E testing framework to verify functionality automatically. If it is a bug, once a repro is provided I'm sure we can iron it out quickly :)

**Comment by aminimalanimal** [comment]
I'm running `plop` via a package.json script: `plop: plop --plopfile config/plop/index.js`.

Version: 3.0.1
Node version: 14.1.0
NPM version: 6.14.14

## config/plop/index.js

```js
import componentGenerator from './component/index.js'

export default function(plop) {
  plop.setGenerator('component', componentGenerator)
  plop.addHelper('curly', (object, open) => (open ? '{' : '}'))
}
```

## config/plop/componentExists.js

Checks whether the given component exists in the components directory

```js
import fs from 'fs'
import path from 'path'

const components = fs.readdirSync(path.join(__dirname, '../../components'))

export default function componentExists(component) {
  return components.includes(component)
}
```

## config/plop/component/index.js

```js
import componentExists from '../componentExists'

export default {
  description: 'Add a Vue component',
  prompts: [
    {
      type: 'input',
      name: 'name',
      message: 'What should it be called?',
      default: 'MyComponent',
      validate: (value) => {
        if ((/.+/).test(value)) {
          return componentExists(value) ? 'A component with this name already exists' : true
        }

        return 'The name is required'
      },
    },
  ],
  actions: [
    {
      type: 'add',
      path: '../../components/{{ properCase name }}/{{ properCase name }}.vue',
      templateFile: './component/component.vue.hbs',
      abortOnFail: true,
    },
  ],
}
```

## config/plop/component/component.vue.hbs
```hbs
<template>
  <div class="component--{{ dashCase name }}">

  </div>
</template>

<script>
export default {
  components: {

  },
  props: {

  },
  data() {
    return {

    }
  },
  computed: {

  },
  methods: {

  },
}

</script>

<style lang="scss">
.component--{{ dashCase name }} {

}

</style>

```

**Comment by aminimalanimal** [comment]
Got it sorted.

This issue had everything to do with my project’s readiness for ES modules. My project can't use `"type": "module"` within the package.json file ('cause it breaks everything else). Updated my node version to 14.18.1 (16 also works for plop, but not for other dependencies in my project).

Some other changes I had to make:
- used the mjs extension instead of js
  - renamed config/plop/index.js to config/plop/index.mjs
  - renamed config/plop/componentExists.js to config/plop/componentExists.mjs
  - renamed config/plop/component/index.js to config/plop/component/index.mjs
  - updated `import componentExists from '../componentExists' to explicitly use the extension
  - updated the package.json plop command to `plop --plopfile config/plop/index.mjs`
- changed the internals of `componentExists` to handle ES Modules not having access to `__dirname` required adding these lines:
  ```js
  import { fileURLToPath } from 'url'

  const __dirname = path.dirname(fileURLToPath(import.meta.url))
  ```

I’m not surprised my stuff wasn't ready for this change; I had no intention of making it. But whatevs, it's done now; may as well keep it.
